### PR TITLE
Ieatkillerbees/ch60/listfamilies endpoint

### PR DESCRIPTION
**Pull request by ieatkillerbees**

### CODE REVIEW

File: src/Actions/Family/ListFamilies.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+use Assert\Assert;
+use Choredo\Entities\Family;
+use Choredo\EntityManagerAwareInterface;
+use Choredo\Filter;
+use Choredo\Filterable;
+use Choredo\HasEntityManager;
+use Choredo\Output\CreatesFractalScope;
+use Choredo\Output\FractalAwareInterface;
+use Choredo\Pageable;
```
Comment (greydnls) [inline src/Actions/Family/ListFamilies.php]: I like to put these into a namespace together, usually something like `Choredo\Actions\Behaviors` or `Choredo\Actions\Contracts` Thoughts?

File: src/Actions/Family/ListFamilies.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    public static function getMaxLimit(): int
+    {
+        return Pageable::DEFAULT_MAX_LIMIT;
+    }
+
+    public static function getDefaultSort(): array
+    {
+        return [
+            ['createdDate', 'ASC'],
```
Comment (greydnls) [inline src/Actions/Family/ListFamilies.php]: I think there might be an ASC constant. http://php.net/manual/en/array.constants.php

File: src/App.php [hunk 3]
```diff
@@ -51,14 +51,16 @@ public function run()
...
-
             }
             $response = new Response\JsonResponse([], 200, [
                 'Access-Control-Allow-Origin' => '*',
                 'Access-Control-Allow-Methods' => 'GET,PUT,POST,DELETE,OPTIONS',
                 'Access-Control-Allow-Headers' => 'Content-Type, Authorization, Content-Length, X-Requested-With',
             ]);
         } catch (\Throwable $e) {
+            if (getenv('APP_ENV') === 'local') {
+                throw $e;
```
Comment (greydnls) [inline src/App.php:62]: as a side note, we should put in Monolog and log some shiz.

File: src/Entities/Behaviors/HasCreatedDate.php [hunk 4]
```diff
@@ -25,7 +25,7 @@ public function updateCreatedDate()
     /**
      * @return \DateTime
      */
-    public function getCreatedDate() :\DateTime
+    public function getCreatedDate(): ?\DateTime
```
Comment (greydnls) [inline src/Entities/Behaviors/HasCreatedDate.php]: Why would this ever return null?

File: src/Pagination.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Choredo;
+
+class Pagination
```
Comment (greydnls) [inline src/Pagination.php]: This is pedantic, but I feel like `Pagination` is a weird name for a DTO. in the past I've used `OffsetLimit`. Thoughts?

File: src/Repositories/FamilyRepository.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+class FamilyRepository extends EntityRepository
+{
+    /**
+     * @param Pagination $pagination
+     * @param Sort[] $sorts
+     * @param Filter[] $filters
+     * @return Paginator
+     */
+    public function getAll(Pagination $pagination, array $sorts = [], array $filters = []): Paginator
+    {
```
Comment (greydnls) [inline src/Repositories/FamilyRepository.php]: Is it worth validating types on `$sorts` and `$filters`, you're calling mthods on them with no guarantee that they're being used correctly.

File: src/Transformers/FamilyTransformer.php [hunk 7]
```diff
@@ -11,13 +11,23 @@ class FamilyTransformer extends TransformerAbstract
...
-            'id'         => $family->getId()->toString(),
+        $entity = [
+            'id'                  => $family->getId()->toString(),
             'name'                => $family->getName(),
             'paymentStrategy'     => $family->getPaymentStrategy(),
             'completionThreshold' => $family->getCompletionThreshold(),
             'weekStartDay'        => $this->transformWeekStartDay($family->getWeekStartDay()),
         ];
+
+        if ($family->getCreatedDate()) {
```
Comment (greydnls) [inline src/Transformers/FamilyTransformer.php]: Why would this ever be null?

File: src/Transformers/FamilyTransformer.php [hunk 8]
```diff
@@ -11,13 +11,23 @@ class FamilyTransformer extends TransformerAbstract
...
+            $entity['createdDate'] = $family->getCreatedDate()->format(\DateTime::ATOM);
+        }
+
+        if ($family->getUpdateDate()) {
```
Comment (greydnls) [inline src/Transformers/FamilyTransformer.php:26]: Let's change the contract back to not-nullable and call the `updateCreatedDate` etc methods explicitly in tests.

**Review by greydnls — CHANGES_REQUESTED** [review]
Dates no null pls. errythang else is my opinion and you're welcome to ignore it as per usual. :xoxo

**Review by greydnls — APPROVED** [review]

File: src/Actions/Family/CreateFamily.php [hunk 9]
```diff
@@ -31,7 +32,8 @@ public function __construct(EntityManagerInterface $entityManager)
 
     public function __invoke(ServerRequestInterface $request, ResponseInterface $response)
     {
-        $family = $request->getAttribute('familyEntity');
+        /** @var \Choredo\Entities\Family $family */
+        $family = $request->getAttribute(REQUEST_RESOURCE);
```
Comment (greydnls) [inline src/Actions/Family/CreateFamily.php:39]: Is it worth asserting that this is what you think it is?

File: src/Actions/Family/ListFamilies.php [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    use Behaviors\HasDefaultCreateDateSort;
+    use Behaviors\HasDefaultPaginationLimits;
+
+    /**
+     * @return array {
+     * @type string $field Name of the field which requires a transform
+     * @type callable $callable A callable transform function `function ($value) {}`
+     * }
+     */
+    public static function getFilterTransforms(): array
```
Comment (greydnls) [inline src/Actions/Family/ListFamilies.php]: RidiculouslyPedantic.gif: Consider you keep interface methods together?

File: src/Container.php [hunk 11]
```diff
@@ -36,5 +37,7 @@ function () {
 
         $this->inflector(FractalAwareInterface::class)
             ->invokeMethod('setManager', [$this->get(Manager::class)]);
+        $this->inflector(EntityManagerAwareInterface::class)
+            ->invokeMethod('setEntityManager', [$this->get(EntityManagerInterface::class)]);
```
Comment (greydnls) [inline src/Container.php]: inflectors are bae.

File: src/Filterable.php [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Choredo;
+
+interface Filterable
```
Comment (greydnls) [inline src/Filterable.php]: I think this is leftover cruft since you moved this already.

**Review by greydnls — CHANGES_REQUESTED** [review]

File: src/Actions/Family/CreateFamily.php (hunk 9 above)
Comment (ieatkillerbees) [reply src/Actions/Family/CreateFamily.php:39]: The inline docblock? I just like to get the sweet, sweet autocomplete in my IDE :o

File: composer.json [hunk 13]
```diff
@@ -16,9 +16,11 @@
         "ramsey/uuid-doctrine": "^1.4",
         "shrikeh/teapot": "^2.3",
         "vlucas/phpdotenv": "~2.2",
-        "zendframework/zend-diactoros": "^1.6"
+        "zendframework/zend-diactoros": "^1.6",
+        "zendframework/zend-paginator": "^2.8"
```
Comment (greydnls) [inline composer.json:20]: I don't think you're using this...

**Review by greydnls — APPROVED** [review]

**Pull request merged**